Code/Framework/AzCore/AzCore/Settings/CommandLineParser.h

```diff
@@ -0,0 +1,173 @@
+
+        //! Callback function which is invoked when an option argument(-short=value or --long=value) is processed
+        //! to split the the option into an option name value pair
+        //! @return Argument containing the option name and Token
+        using OptionDelimiterFunc = AZStd::function<AZStd::optional<CommandLineArgument>(AZStd::string_view token)>;
+
+        //! Function which is invoked to split an option into an option name and option value pair
+        OptionDelimiterFunc m_optionDelimiterFunc = &DefaultOptionDelimiter;
+
+        //! By default the option prefixes are "--" and "-"
+        //! NOTE: The order is important here as double dash is checked first before single dash
+        //! to make sure an argument such as `--foo bar` parses both dashes before the option name.
+        //! Otherwise the option name would be parses as "-foo" which is incorrect.
+        //! The correct parsing is "foo" for the option name
+        CommandLineOptionPrefixArray m_optionPrefixes{ Platform::GetDefaultOptionPrefixes() };
+
+        //! Separator which is used to separate normal argument parsing(option/positional) from just positional argument parsing
+        //! Positional arguments can still be parsed before the separator, however options will not be parsed
+        //! (i.e `-- --foo bar`) will parse `--foo` as a positional argument not an option with a value of "bar"
+        AZStd::string_view m_positionalArgSeparator = "--";
+    };
+
+    // Expected structure which encapsulates the result of the ParseCommandLine function
+    using CommandLineParseErrorString = AZStd::basic_fixed_string<char, 512, AZStd::char_traits<char>>;
+    using CommandLineParseOutcome = AZStd::expected<void, CommandLineParseErrorString>;
+
+    //! Parses a set of command line arguments
+    //! This function will NOT allocate any heap memory on its on
+    //! It is safe to call without any AZ Allocators being initialized
+    //!
+    //! @param commandLine A contiguous range of string_view arguments pointing to each command line token
+    //! @param configParseSettings struct defining how the command line arguments should be parsed
+    //! @return success outcome if the each token on the command line was parsed successfully,
+    //! otherwise a failure string with the parse error is returned
+    CommandLineParseOutcome ParseCommandLine(
+        AZStd::span<AZStd::string_view const> commandLine, const CommandLineParserSettings& commandLineParserSettings);
+
+    //! Parses a set of command line arguments using the defacto integer argument count parameter and char double pointer to the argument
+    //! array
+    //! @param argc number of command line arguments
+    //! @param argv pointer to the begining of array of c-strings pointing to each command line argument
+    CommandLineParseOutcome ParseCommandLine(int argc, char** argv, const CommandLineParserSettings& commandLineParserSettings);
```

Comment:
The first one of these treats the executable name (arg0) as a parsed setting, and the second one doesn't, right? If so, this should get called out in the comments.

Reply:
I think you are confused with existing [CommandLine::Parse](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Settings/CommandLine.cpp#L185-L212) function which when passing in an integer argument count and a c-string array argument value pointer skips the first argument.

Both of these `ParseCommandLine` functions, doesn't treat arg0 specially and parses it normally.

Reply:
Whoops, yep, thanks